routes/main.js

```diff
@@ -0,0 +1,219 @@
+
+//     product.save((err) => {
+//       if (err) throw err;
+//     });
+//   }
+//   res.end();
+// });
+
+router.param("product", function (req, res, next, id) {
+  Product.find({ _id: id }, function (err, product) {
+    if (err) {
+      next(err);
+    } else {
+      req.product = product[0];
+      next();
+    }
+  });
+});
+
+router.param("review", function (req, res, next, id) {
+  Review.find({ _id: id }, function (err, review) {
+    if (err) {
+      next(err);
+    } else {
+      req.review = review[0];
+      next();
+    }
+  });
+});
+
+router.get("/products", (req, res, next) => {
+  const perPage = 9;
+  const page = req.query.page || 1;
+
+  let query;
+  let sortData;
+  let data = {};
+
+  if (!req.query) {
+    query = {};
+  }
+  //category
+  if (req.query.category) {
+    query = { category: req.query.category.toUpperCase() };
+  }
+
+  //sort by price
+  if (req.query.price) {
+    if (req.query.price === "highest") {
+      sortData = { price: -1 };
+    } else if (req.query.price === "lowest") {
+      sortData = { price: 1 };
+    } else {
+      sortData = null;
+    }
+  }
+
+  //db search based on user search input
+  if (req.query.query) {
+    const searchString = req.query.query.toUpperCase();
+
+    //checks to see if string = 1 word
+    //if 1 word search all docs for that word with or w/o category
+    if (searchString.split(/\W+/).length === 1) {
+      if (req.query.query && req.query.category) {
+        query = {
+          $text: { $search: `${searchString}` },
+          category: req.query.category.toUpperCase(),
+        };
+      } else {
+        query = { $text: { $search: searchString } };
+      }
+      //if > 1 word do exact match w/ or w/o category
+    } else {
+      if (req.query.query && req.query.category) {
+        query = {
+          $text: { $search: `\"${searchString}\"` },
+          category: req.query.category.toUpperCase(),
+        };
+      } else {
+        query = {
+          $text: { $search: `\"${searchString}\"` },
+        };
+      }
+    }
+  }
+
+  Product.find(query)
+    .sort(sortData)
+    .skip(perPage * page - perPage)
+    .limit(perPage)
+    .exec((err, products) => {
+      Product.countDocuments(query, (err, count) => {
+        if (err) return next(err);
+
+        data = {
+          products: products,
+          count: count,
+        };
+
+        if (data.count === 0) {
+          res.status(404).send("Product not found.");
+        } else {
+          res.status(200).send(data);
+        }
+      });
+    });
+});
+
+router.get("/products/:product", (req, res, next) => {
+  if (!req.product) {
+    res.status(404).send("Product not found.");
+  } else {
+    res.status(200).send(req.product);
+  }
+
+  //const id = req.params.product;
+  // Product.findById(id).exec((err, product) => {
+  //   if (err) return next(err);
+  //   res.status(200).send(product);
+  //   res.end();
+  // });
```

Comment:
Usually you'd want to remove commented out code like those or comment as to why you left it in